### PR TITLE
HIVE-24330: Automate setting permissions on cmRoot directories.

### PR DESCRIPTION
**Pull request by ArkoSharma**

### CODE REVIEW

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/ReplChangeManager.java [hunk 1]
```diff
@@ -595,6 +604,42 @@ public Void execute() throws IOException {
         if (!cmFs.exists(cmroot)) {
           cmFs.mkdirs(cmroot);
           cmFs.setPermission(cmroot, new FsPermission("700"));
+          //add additional acls to ensure that owner and group of warehouse have access.
+          try {
+            FileStatus warehouseStatus = cmFs.getFileStatus(new Path(MetastoreConf.get(conf, ConfVars.WAREHOUSE.getVarname())));
```
Comment (aasha) [inline standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/ReplChangeManager.java:609]: Create separate method

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/ReplChangeManager.java [hunk 2]
```diff
@@ -20,21 +20,25 @@
...
+import com.google.common.collect.Lists;
 import org.apache.commons.lang3.concurrent.BasicThreadFactory;
 import org.apache.hadoop.conf.Configuration;
-import org.apache.hadoop.fs.FileChecksum;
-import org.apache.hadoop.fs.FileStatus;
-import org.apache.hadoop.fs.FileSystem;
-import org.apache.hadoop.fs.Path;
-import org.apache.hadoop.fs.PathFilter;
-import org.apache.hadoop.fs.Trash;
+import org.apache.hadoop.fs.*;
```
Comment (aasha) [inline standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/ReplChangeManager.java:35]: Fix this import

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/ReplChangeManager.java [hunk 3]
```diff
@@ -606,6 +610,44 @@ public Void execute() throws IOException {
...
+   * Assign cmRoot to group of warehouse directory.
+   * If warehouse directory cannot be determined then give rwx permissions to default group of cmroot.
+   */
+  private static void setCmRootPermissions(Path cmroot) throws IOException{
+    FileSystem cmFs = cmroot.getFileSystem(conf);
+    cmFs.setPermission(cmroot, new FsPermission("770"));
+    try {
+      FileStatus warehouseStatus = cmFs.getFileStatus(new Path(MetastoreConf.get(conf, ConfVars.WAREHOUSE.getVarname())));
+
+      //if wh-owner is same as cmroot owner, then change the group if required.
```
Comment (aasha) [inline standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/ReplChangeManager.java:623]: this comment is not clear

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/ReplChangeManager.java [hunk 4]
```diff
@@ -606,6 +610,44 @@ public Void execute() throws IOException {
...
+      if (warehouseOwner.equals(cmFs.getFileStatus(cmroot).getOwner())) {
+        FsAction whOwnerAction = warehouseStatus.getPermission().getUserAction();
+        FsAction whGroupAction = warehouseStatus.getPermission().getGroupAction();
+        FsAction whOtherAction = warehouseStatus.getPermission().getOtherAction();
+        if(!warehouseGroup.equals(cmFs.getFileStatus(cmroot).getGroup())) {
+          cmFs.setOwner(cmroot, null, warehouseGroup);
+          cmFs.setPermission(cmroot, new FsPermission(whOwnerAction, whGroupAction, whOtherAction));
+        }
+      } else {
+        if(!warehouseGroup.equals(cmFs.getFileStatus(cmroot).getGroup())) {
```
Comment (aasha) [inline standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/ReplChangeManager.java:636]: A log line here stating cmroot and warehouse owners are different

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/metastore/TestReplChangeManager.java [hunk 5]
```diff
@@ -434,6 +437,60 @@ public void testRecycleUsingImpersonation() throws Exception {
     } while (!cleared);
   }
 
+  @Test
+  public void tesRecyleImpersionationWithGroupPermissions() throws Exception {
```
Comment (aasha) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/metastore/TestReplChangeManager.java:441]: test for acl access

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/ReplChangeManager.java [hunk 6]
```diff
@@ -606,6 +614,45 @@ public Void execute() throws IOException {
...
+      } else {
+        LOG.warn("Metastore-user is not same as owner of warehouse.");
+        if(!warehouseGroup.equals(cmFs.getFileStatus(cmroot).getGroup())) {
+          List<AclEntry> aclList = Lists.newArrayList(
+                  new AclEntry.Builder().setScope(ACCESS).setType(USER).setPermission(FsAction.ALL).build(),
+                  new AclEntry.Builder().setScope(ACCESS).setType(GROUP).setPermission(FsAction.ALL).build(),
+                  new AclEntry.Builder().setScope(ACCESS).setType(OTHER).setPermission(FsAction.NONE).build());
+          aclList.add(new AclEntry.Builder().setScope(ACCESS).setType(GROUP).setName(warehouseGroup).
+                  setPermission(warehouseStatus.getPermission().getGroupAction()).build());
+          cmFs.setAcl(cmroot, aclList);
```
Comment (aasha) [inline standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/ReplChangeManager.java:648]: If group is same and owner is different for cm and warehouse nothing needs to be done?

Comment (ArkoSharma) [reply standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/ReplChangeManager.java:648]: Group permissions were provided initially(line 624), so nothing needs to be added.

**Pull request closed without merging**